src/backend/gpopt/translate/CQueryMutators.cpp

```diff
@@ -35,6 +35,494 @@ extern "C" {
 using namespace gpdxl;
 using namespace gpmd;
 
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CQueryMutators::GroupingListContainsPrimaryKey
+//
+//	@doc:
+//		Checks if list grouping_list contains all columns from conkeys of
+//		Primary Key in relation conrelid. grouping_list is treated as flat
+//		list of SortGroupClauses, other nodes are ignored.
+//---------------------------------------------------------------------------
+BOOL
+CQueryMutators::GroupingListContainsPrimaryKey(Query *query,
+											   List *grouping_list,
+											   Bitmapset *conkeys, Oid conrelid)
+{
+	BOOL result = false;
+	Bitmapset *grouping_set = NULL;
+	ListCell *lgc;
+
+	// Check if conkeys set is a subset of grouping columns set.
+	ForEach(lgc, grouping_list)
+	{
+		Node *grpcl = (Node *) lfirst(lgc);
+
+		if (NULL == grpcl || !IsA(grpcl, SortGroupClause))
+		{
+			continue;
+		}
+
+		SortGroupClause *sgc = (SortGroupClause *) grpcl;
+		Node *expr = gpdb::GetSortGroupClauseExpr(sgc, query->targetList);
+
+		GPOS_ASSERT(IsA(expr, Var));
+
+		Var *var = (Var *) expr;
+
+		RangeTblEntry *rte =
+			(RangeTblEntry *) gpdb::ListNth(query->rtable, var->varno - 1);
+
+		GPOS_ASSERT(NULL != rte);
+
+		if (0 == var->varlevelsup && rte->relid == conrelid)
+		{
+			grouping_set = gpdb::BmsAddMember(grouping_set, var->varattno);
+		}
+	}
+
+	result = gpdb::BmsIsSubset(conkeys, grouping_set);
```

Comment:
```suggestion
	BOOL result = gpdb::BmsIsSubset(conkeys, grouping_set);
```
or even move BmsIsSubset out of this function

Reply:
Then we'll need to pull ```BmsFree``` as well... Do not see much reason for that.